nerfstudio/generative/deepfloyd.py

```diff
@@ -0,0 +1,207 @@
+import sys
+import gc 
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional, Union
+
+import appdirs
+import mediapy
+import numpy as np
+import torch
+import torch.nn.functional as F
+import tyro
+from rich.console import Console
+from torch import nn
+from torch.cuda.amp import custom_bwd, custom_fwd
+from torch.cuda.amp.grad_scaler import GradScaler
+from torchtyping import TensorType
+
+from diffusers import IFPipeline
+from transformers import T5EncoderModel, T5Tokenizer
+
+IMG_DIM = 64
+
+class _SDSGradient(torch.autograd.Function):  # pylint: disable=abstract-method
+    """Custom gradient function for SDS loss. Since it is already computed, we can just return it."""
+
+    @staticmethod
+    @custom_fwd
+    def forward(ctx, input_tensor, gt_grad):  # pylint: disable=arguments-differ
+        del input_tensor
+        ctx.save_for_backward(gt_grad)
+        # Return magniture of gradient, not the actual loss.
+        return torch.mean(gt_grad**2) ** 0.5
+
+    @staticmethod
+    @custom_bwd
+    def backward(ctx, grad):  # pylint: disable=arguments-differ
+        del grad
+        (gt_grad,) = ctx.saved_tensors
+        batch_size = len(gt_grad)
+        return gt_grad / batch_size, None
```

Comment:
Create a new utils folder in `generative/` that has this so that It can be shared between SD and deepfloyd